File: Coding_Questions/Python/GroupAnagrams.py
# ---Group Anagrams---

from typing import List


def group_anagrams(words: List[str]) -> List[List[str]]:
    return_list = list()
    used_words = dict()

    for i in range(len(words)):
        current_word = words[i]
        anagrams_list = [current_word]

        if words[i] in used_words.keys():
            continue

        for r in range(len(words)):
            if i == r:
                continue

            if are_anagrams(words[i], words[r]):
                anagrams_list.append(words[r])
                used_words[words[r]] = 0

        return_list.append(anagrams_list)

    return return_list


def are_anagrams(word1: str, word2: str):
    return sorted(word1) == sorted(word2)


words_input = ["yo", "act", "flop", "tac", "foo", "cat", "oy", "olfp"]
expected = [["yo", "oy"], ["flop", "olfp"], ["act", "tac", "cat"], ["foo"]]
output = list(map(lambda x: sorted(x), group_anagrams(words_input)))

print(output)
# [["yo", "oy"], ["flop", "olfp"], ["act", "tac", "cat"], ["foo"]]

File: Coding_Questions/C++/GroupAnagrams.cpp
// ---Group Anagrams---

#include <vector>
#include <map>


bool are_anagrams(std::string word1, std::string word2) {
    std::sort(word1.begin(), word1.end());
    std::sort(word2.begin(), word2.end());
    return word1 == word2;
}


std::vector<std::vector<std::string>> group_anagrams(std::vector<std::string> words) {
    std::vector<std::vector<std::string>> return_vector;
    std::map<std::string, size_t> words_map;

    for (size_t i = 0; i < words.size(); ++i) {
        std::vector<std::string> current_vector = {words[i]};
        if (words_map.find(words[i]) != words_map.cend()) {
            continue;
        }
        for (size_t r = 0; r < words.size(); ++r) {
            if (r == i) {
                continue;
            }

            if (are_anagrams(words[i], words[r]) && words_map.find(words[r]) == words_map.cend()) {
                current_vector.push_back(words[r]);
                words_map[words[r]] = 0;
            }
        }
        return_vector.push_back(current_vector);
    }

    return return_vector;
}